CodeJam.Main/Threading/TaskHelper.ForEachAsync.cs

```diff
@@ -0,0 +1,191 @@
+#else
+using TaskEx = System.Threading.Tasks.TaskEx;
+#endif
+
+namespace CodeJam.Threading
+{
+	/// <summary>
+	/// Helper methods for <see cref="Task"/> and <see cref="Task{TResult}"/>.
+	/// </summary>
+	public static partial class TaskHelper
+	{
+		private const int _maxProcessorCountRefreshTicks = 30000;
+		private static volatile int _processorCount;
+		private static volatile int _lastProcessorCountRefreshTicks;
+
+		// BASEDON PlatformHelper.ProcessorCount
+		private static int ProcessorCount
+		{
+			get
+			{
+				var tickCount = Environment.TickCount;
+				if (_processorCount == 0 || tickCount - _lastProcessorCountRefreshTicks >= _maxProcessorCountRefreshTicks)
+				{
+					_processorCount = Environment.ProcessorCount;
+					_lastProcessorCountRefreshTicks = tickCount;
+				}
+				return _processorCount;
+			}
+		}
+
+		/// <summary>
+		/// Gets the maximum degree of parallelism for the scheduler.
+		/// Matches to the <see cref="Parallel.ForEach{TSource}(IEnumerable{TSource}, Action{TSource})"/> behavior.
+		/// Limits <paramref name="value"/> by <see cref="TaskScheduler.MaximumConcurrencyLevel"/> value (if non-zero positiver value).
+		/// Otherwise, uses <see cref="Environment.ProcessorCount"/> as fallback value.
+		/// </summary>
+		public static int GetMaxDegreeOfParallelism([NotNull] this TaskScheduler scheduler, int value)
+		{
+			var concurrencyLimit = scheduler.MaximumConcurrencyLevel;
+
+			if (concurrencyLimit > 0 && concurrencyLimit != int.MaxValue)
+				return value <= 0 ? concurrencyLimit : Math.Min(concurrencyLimit, value);
+
+			return value <= 0 ? ProcessorCount : value;
+		}
+
+		/// <summary>
+		/// Runs actions over source items concurrently and asynchronously.
+		/// </summary>
+		/// <typeparam name="T">Type of items to process</typeparam>
+		/// <param name="source">The source.</param>
+		/// <param name="callback">The callback.</param>
+		/// <param name="maxDegreeOfParallelism">
+		/// The maximum degree of parallelism. If zero or negative, default scheduler value is used.
+		/// See <see cref="GetMaxDegreeOfParallelism"/> documentation for more details.
+		/// </param>
+		/// <param name="cancellation">The cancellation.</param>
+		// BASEDON https://stackoverflow.com/a/25877042
+		public static async Task ForEachAsync<T>(
+			[NotNull] this IEnumerable<T> source,
+			[NotNull] Func<T, CancellationToken, Task> callback,
+			int maxDegreeOfParallelism = 0,
+			CancellationToken cancellation = default)
+		{
+			maxDegreeOfParallelism = TaskScheduler.Current.GetMaxDegreeOfParallelism(maxDegreeOfParallelism);
+
+			using (var failureCancellation = new CancellationTokenSource())
+			using (var combinedCancellation = CreateCancellation(cancellation, failureCancellation.Token))
+			{
+				var combinedToken = combinedCancellation.Token;
+
+				await TaskEx
+					.WhenAll(
+						from partition in Partitioner.Create(source).GetPartitions(maxDegreeOfParallelism)
+						select TaskEx.Run(
```

Comment:
I don't see how LINQ syntax helps here.

```suggestion
Partitioner.Create(source).GetPartitions(maxDegreeOfParallelism).Select(partition =>
						TaskEx.Run(
```